test/rules/implicit_reopen_test.dart

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test_reflective_loader/test_reflective_loader.dart';
+
+import '../rule_test_support.dart';
+
+main() {
+  defineReflectiveSuite(() {
+    defineReflectiveTests(ImplicitReopenTest);
+  });
+}
+
+@reflectiveTest
+class ImplicitReopenTest extends LintRuleTest {
+  @override
+  bool get addMetaPackageDep => true;
+
+  @override
+  List<String> get experiments => ['class-modifiers', 'sealed-class'];
+
+  @override
+  String get lintRule => 'implicit_reopen';
+
+  test_extends_class_classFinal_ok() async {
+    await assertNoDiagnostics(r'''
+final class F {}
+
+class C extends F {}
+''');
+  }
+
+  test_extends_class_classInterface() async {
+    await assertDiagnostics(r'''
+interface class I {}
+
+class C extends I {}
+''', [
+      lint(28, 1),
+    ]);
+  }
+
+  test_extends_class_classInterface_outsideLib_ok() async {
+    newFile('$testPackageLibPath/a.dart', r'''
+interface class I {}
+''');
+
+    await assertDiagnostics(r'''
+import 'a.dart';
+
+class C extends I {}
+''', [
+      error(CompileTimeErrorCode.INTERFACE_CLASS_EXTENDED_OUTSIDE_OF_LIBRARY,
+          34, 1),
+    ]);
+  }
+
+  test_extends_class_classInterface_reopened_ok() async {
+    await assertNoDiagnostics(r'''
+import 'package:meta/meta.dart';
+
+interface class I {}
+
+@reopen
+class C extends I {}
+''');
+  }
+
+  test_extends_class_classSealed_classInterface() async {
+    await assertDiagnostics(r'''
+interface class I {}
+
+sealed class S extends I {}
```

Comment:
Okay, I was rereading Erik and Lasse's comments after looking at these tests, and I am also confused if there's a need to go up the class hierarchy to find implicitly induced modifiers. Is it solely just looking at the `sealed` type's direct supertypes (extends, with)?

Would the following test case with another level of `sealed` not also give a lint @eernstg @lrhn? (Also, if that's the case, I'm sorry @pq to have misled you D:)
```dart
interface class I {}
sealed class S extends I {}
sealed class AnotherS extends S {}
class C extends S {}
```

Also, may be a good test case?

Reply:
> if that's the case, I'm sorry @pq to have misled you D:

No worries!  I'm just glad to get some early feedback so we can get this right.

Reply:
Thanks!